File: packages/schemas/rust/src/prompt.rs
//! Prompt metadata bindings (closed record).

use serde::{Deserialize, Serialize};

use crate::schema_version::SchemaVersionV1;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PromptStage {
    Planner,
    Search,
    Extraction,
    Entity,
    Relation,
    Validation,
    Writing,
    Synthesis,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ModelCapability {
    Strong,
    Medium,
    Cheap,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SchemaRef {
    pub schema: String,
    pub schema_version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ModelHint {
    pub capability: ModelCapability,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct GoldenCase {
    pub case_id: String,
    pub fixture_ref: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PromptMetadata {
    pub schema_version: SchemaVersionV1,
    pub prompt_id: String,
    pub version: String,
    pub stage: PromptStage,
    pub purpose: String,
    pub input_schema: SchemaRef,
    pub output_schema: SchemaRef,
    pub model_hint: Option<ModelHint>,
    pub safety: Option<Vec<String>>,
    pub golden_cases: Vec<GoldenCase>,
}

use crate::ContractValid;

impl ContractValid for SchemaRef {
    fn check_contract(&self) -> Result<(), String> {
        if self.schema.len() < 3 {
            return Err("schema reference name must be at least 3 characters".into());
        }
        if self.schema_version.is_empty() {
            return Err("schema reference version must not be empty".into());
        }
        Ok(())
    }
}

impl ContractValid for GoldenCase {
    fn check_contract(&self) -> Result<(), String> {
        if self.case_id.len() < 3 {
            return Err("case_id must be at least 3 characters".into());
        }
        if self.fixture_ref.len() < 3 {
            return Err("fixture_ref must be at least 3 characters".into());
        }
        Ok(())
    }
}

impl ContractValid for PromptMetadata {
    fn check_contract(&self) -> Result<(), String> {
        if self.prompt_id.len() < 5 {
            return Err("prompt_id must be at least 5 characters".into());
        }
        if self.version.len() < 5 {
            return Err("version must be a full SemVer string (e.g. '1.0.0')".into());
        }
        if self.purpose.is_empty() {
            return Err("purpose must not be empty".into());
        }
        self.input_schema.check_contract()?;
        self.output_schema.check_contract()?;
        if self
            .safety
            .as_ref()
            .is_some_and(|items| items.iter().any(|s| s.is_empty()))
        {
            return Err("safety entries must not be empty".into());
        }
        if self.golden_cases.is_empty() {
            return Err("golden_cases must have at least one entry".into());
        }
        for case in &self.golden_cases {
            case.check_contract()?;
        }
        Ok(())
    }
}

File: packages/schemas/ts/src/prompt.ts
import { z } from "zod";
import { SchemaVersionV1 } from "./schema-version";

/** Prompt metadata is a closed record: unknown fields are rejected in every stack. */
export const PromptStage = z.enum([
  "planner",
  "search",
  "extraction",
  "entity",
  "relation",
  "validation",
  "writing",
  "synthesis",
]);

export const ModelCapability = z.enum(["strong", "medium", "cheap"]);

const SchemaRefSchema = z
  .object({
    schema: z.string().min(3),
    schema_version: z.string().min(1),
  })
  .strict();
export type SchemaRef = z.infer<typeof SchemaRefSchema>;

export const GoldenCaseSchema = z
  .object({
    case_id: z.string().min(3),
    fixture_ref: z.string().min(3),
  })
  .strict();
export type GoldenCase = z.infer<typeof GoldenCaseSchema>;

export const PromptMetadataSchema = z
  .object({
    schema_version: SchemaVersionV1,
    prompt_id: z.string().min(5),
    version: z.string().min(5),
    stage: PromptStage,
    purpose: z.string().min(1),
    input_schema: SchemaRefSchema,
    output_schema: SchemaRefSchema,
    model_hint: z
      .object({
        capability: ModelCapability,
      })
      .strict()
      .optional(),
    safety: z.array(z.string().min(1)).optional(),
    golden_cases: z.array(GoldenCaseSchema).min(1),
  })
  .strict();
export type PromptMetadata = z.infer<typeof PromptMetadataSchema>;
